File: src/main/java/bwdm/boundaryValueAnalysisUnit/BoundaryValueAnalyzer.java
package bwdm.boundaryValueAnalysisUnit;

import bwdm.informationStore.InformationExtractor;
import bwdm.Util;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.stream.Collectors;

public class BoundaryValueAnalyzer {

	static final long intMax  = Integer.MAX_VALUE;
	static final long intMin  = Integer.MIN_VALUE;
	static final long natMax  = intMax * 2;
	static final long natMin  = 0;
	static final long nat1Max = natMax + 1;
	static final long nat1Min = 1;

	private HashMap<String, ArrayList<Long>> boundaryValueList;
	private ArrayList<HashMap<String, Long>> inputDataList;


	public BoundaryValueAnalyzer(InformationExtractor _information) {
		//generation of instance of each parameter
		boundaryValueList = new HashMap<>();
		inputDataList = new ArrayList<>();
		_information.getParameters().forEach(p -> boundaryValueList.put(p, new ArrayList<>()));

		generateTypeBoundaryValue(_information);
		generateIfConditionalBoundaryValue(_information);

		//remove overlapped values
		ArrayList<String> parameters = _information.getParameters();
		for(int i = 0; i < boundaryValueList.size(); i++) {
			String parameter = parameters.get(i);
			ArrayList<Long> bvs = boundaryValueList.get(parameter);
			bvs = (ArrayList<Long>) bvs.stream().distinct().collect(Collectors.toList());
			boundaryValueList.put(parameter, bvs);
		}

		makeInputDataList(_information);
	}

	public HashMap<String, ArrayList<Long>> getBoundaryValueList() { return boundaryValueList; }
	public ArrayList<HashMap<String, Long>> getInputDataList() { return inputDataList; }


	private void generateTypeBoundaryValue(InformationExtractor _information) {
		ArrayList<String> parameters = _information.getParameters();
		ArrayList<String> argumentTypes = _information.getArgumentTypes();

		for(int i=0; i<argumentTypes.size(); i++) {
			String parameter = parameters.get(i);
			String argumentType = argumentTypes.get(i);
			ArrayList<Long> bvs = boundaryValueList.get(parameter);
			switch (argumentType) {
				case "int":
					bvs.add(intMax + 1);
					bvs.add(intMax);
					bvs.add(intMin);
					bvs.add(intMin - 1);
					break;
				case "nat":
					bvs.add(natMax + 1);
					bvs.add(natMax);
					bvs.add(natMin);
					bvs.add(natMin - 1);
					break;
				case "nat1":
					bvs.add(nat1Max + 1);
					bvs.add(nat1Max);
					bvs.add(nat1Min);
					bvs.add(nat1Min - 1);
					break;
				default:
					break;
			}
		}
	}


	private void generateIfConditionalBoundaryValue(InformationExtractor _information) {
		HashMap<String, ArrayList<HashMap<String, String>>> allIfConditions = _information.getIfConditions();

		allIfConditions.forEach( (parameter, ifConditions) ->
                (ifConditions).forEach(condition -> {
                    //condition : HashMap<String, String>
                    String left = condition.get("left");
                    String operator = condition.get("operator");
                    String right = condition.get("right");
                    ArrayList<Long> bvs = boundaryValueList.get(parameter);

                    long trueValue = 0, falseValue = 0, value;
                    if (Util.isNumber(left)) {
                        value = Long.parseLong(left);
                        switch (operator) {
                            case "<":
                                trueValue = value + 1;
                                falseValue = value;
                                break;
                            case "<=":
                                trueValue = value;
                                falseValue = value - 1;
                                break;
                            case ">":
                                trueValue = value - 1;
                                falseValue = value;
                                break;
                            case ">=":
                                trueValue = value;
                                falseValue = value + 1;
                                break;
                            case "mod":
                                trueValue = value + Long.parseLong(condition.get("surplus"));
                                falseValue = value + 1;
                                bvs.add(value - 1);
                                break;
                        }

                    } else if (Util.isNumber(right)) {
                        value = Long.parseLong(right);
                        switch (operator) {
                            case "<":
                                trueValue = value - 1;
                                falseValue = value;
                                break;
                            case "<=":
                                trueValue = value;
                                falseValue = value + 1;
                                break;
                            case ">":
                                trueValue = value + 1;
                                falseValue = value;
                                break;
                            case ">=":
                                trueValue = value;
                                falseValue = value - 1;
                                break;
                            case "mod":
                                trueValue = value + Long.parseLong(condition.get("surplus"));
                                falseValue = value + 1;
                                bvs.add(value - 1);
                                break;
                        }
                    }

                    bvs.add(trueValue);
                    bvs.add(falseValue);

                }));
	}

	private void makeInputDataList(InformationExtractor _information) {
		ArrayList<String> parameters = _information.getParameters();

		//最初の一つ目
		String first_prm = parameters.get(0);
		ArrayList<Long> first_bvs = boundaryValueList.get(first_prm);
		for(int i=0; i<first_bvs.size(); i++) {
			inputDataList.add(new HashMap<>());
			HashMap<String, Long> hm = inputDataList.get(i);
			hm.put(first_prm, first_bvs.get(i));
		}

		//それ以降
		parameters.forEach(p -> {
			if( !p.equals(first_prm) ) { //最初の要素以外に対して
				ArrayList<Long> current_bvs = boundaryValueList.get(p);

				//inputDataListの第一引数のみを登録した状態
				ArrayList<HashMap<String, Long>> inputDataListInitialState = new ArrayList<>(inputDataList);

				for(int i=0; i<current_bvs.size() - 1; i++) {
					ArrayList<HashMap<String, Long>> inputDataListTmp = new ArrayList<>();
					inputDataListInitialState.forEach(inputDataOriginal -> {
						//inputDataを複製
						HashMap<String, Long> inputData = new HashMap<>();
						inputDataOriginal.forEach(inputData::put);
						inputDataListTmp.add(inputData);
					});
					inputDataList.addAll(inputDataListTmp);
				}
				int repeatTimesOfInsert = 0;
				for (Object current_bv : current_bvs) {
					long currentBv = (long) current_bv;
					int offset = repeatTimesOfInsert * inputDataListInitialState.size();
					for (int k = 0; k < inputDataListInitialState.size(); k++) {
						HashMap<String, Long> inputData = inputDataList.get(k + offset);
						inputData.put(p, currentBv);
					}
					repeatTimesOfInsert++;
				}

			}
		});
	} //end makeInputDatList

}

File: src/main/kotlin/bwdm/boundaryValueAnalysisUnit/BoundaryValueAnalyzer.kt
package bwdm.boundaryValueAnalysisUnit

//import com.github.korosuke613.pict-java
import bwdm.Util
import bwdm.informationStore.InformationExtractor
import java.util.*
import java.util.stream.Collectors
import com.github.korosuke613.pict4java.Factor
import com.github.korosuke613.pict4java.Model
import com.github.korosuke613.pict4java.Pict

typealias BoundaryValueList = HashMap<String, ArrayList<Long>>
typealias InputDataList = ArrayList<HashMap<String, Long>>

class BoundaryValueAnalyzer(_information: InformationExtractor, isPairwise: Boolean = true) {

    val boundaryValueList: BoundaryValueList = HashMap()
    val inputDataList: InputDataList = ArrayList()

    init {
        //generation of instance of each parameter
        _information.parameters.forEach { p -> boundaryValueList[p] = ArrayList() }

        generateTypeBoundaryValue(_information)
        generateIfConditionalBoundaryValue(_information)

        //remove overlapped values
        val parameters = _information.parameters
        for (i in 0 until boundaryValueList.size) {
            val parameter = parameters[i]
            var bvs = boundaryValueList[parameter]!!
            bvs = bvs.stream().distinct().collect(Collectors.toList()) as ArrayList<Long>

            boundaryValueList[parameter] = bvs
        }
        if(isPairwise) {
            makeInputDataListWithPairwise(_information)
        }else{
            makeInputDataList(_information)
        }
    }


    private fun generateTypeBoundaryValue(_information: InformationExtractor) {
        val parameters = _information.parameters
        val argumentTypes = _information.argumentTypes

        for (i in argumentTypes.indices) {
            val parameter = parameters[i]
            val argumentType = argumentTypes[i]
            val bvs = boundaryValueList[parameter]
            when (argumentType) {
                "int" -> {
                    bvs!!.run {
                        add(intMax + 1)
                        add(intMax)
                        add(intMin)
                        add(intMin - 1)
                    }
                }
                "nat" -> {
                    bvs!!.run {
                        add(natMax + 1)
                        add(natMax)
                        add(natMin)
                        add(natMin - 1)
                    }
                }
                "nat1" -> {
                    bvs!!.run {
                        add(nat1Max + 1)
                        add(nat1Max)
                        add(nat1Min)
                        add(nat1Min - 1)
                    }
                }
                else -> {
                }
            }
        }
    }


    private fun generateIfConditionalBoundaryValue(_information: InformationExtractor) {
        val allIfConditions = _information.ifConditions

        allIfConditions.forEach { parameter, ifConditions ->
            ifConditions.forEach { condition ->
                //condition : HashMap<String, String>
                val left = condition["left"]
                if (Util.getOperator(left!!) == "+") {
                    return@forEach
                }
                val operator = condition["operator"]
                val right = condition["right"]
                val bvs = boundaryValueList[parameter]

                var trueValue: Long = 0
                var falseValue: Long = 0
                val value: Long
                if (Util.isNumber(left!!)) {
                    value = java.lang.Long.parseLong(left)
                    when (operator) {
                        "<" -> {
                            trueValue = value + 1
                            falseValue = value
                        }
                        "<=" -> {
                            trueValue = value
                            falseValue = value - 1
                        }
                        ">" -> {
                            trueValue = value - 1
                            falseValue = value
                        }
                        ">=" -> {
                            trueValue = value
                            falseValue = value + 1
                        }
                        "mod" -> {
                            trueValue = value + java.lang.Long.parseLong(condition["surplus"])
                            falseValue = value + 1
                            bvs!!.add(value - 1)
                        }
                    }

                } else if (Util.isNumber(right!!)) {
                    value = java.lang.Long.parseLong(right)
                    when (operator) {
                        "<" -> {
                            trueValue = value - 1
                            falseValue = value
                        }
                        "<=" -> {
                            trueValue = value
                            falseValue = value + 1
                        }
                        ">" -> {
                            trueValue = value + 1
                            falseValue = value
                        }
                        ">=" -> {
                            trueValue = value
                            falseValue = value - 1
                        }
                        "mod" -> {
                            trueValue = value + java.lang.Long.parseLong(condition["surplus"])
                            falseValue = value + 1
                            bvs!!.add(value - 1)
                        }
                    }
                }

                bvs!!.run {
                    add(trueValue)
                    add(falseValue)
                }

            }
        }
    }

    private fun makeInputDataListWithPairwise(_information: InformationExtractor){
        val pict = Pict()
        val model = Model()
        // 因子の取得
        val parameters = _information.parameters

        // ファクターの追加
        for (prm in parameters){
            val bvs = boundaryValueList[prm]
            val factor = Factor(named_level = bvs!!.map { it.toString() }, name = prm)
            model.addFactor(factor)
        }

        // ペアワイズ分析した結果を生成
        pict.setRootModel(model)
        val tests = pict.generate()

        for (test in tests) {
            val hash = HashMap<String, Long>()
            for((index, param) in test.withIndex()){
                hash[model.factors[index].name] = param!!.toLong()
            }
            inputDataList.add(hash)
        }
    }

    private fun makeInputDataList(_information: InformationExtractor) {
        val parameters = _information.parameters

        //最初の一つ目
        val firstPrm = parameters[0]
        val firstBvs = boundaryValueList[firstPrm]
        for (i in firstBvs!!.indices) {
            inputDataList.add(HashMap())
            val hm = inputDataList[i]
            hm[firstPrm] = firstBvs[i]
        }

        //それ以降
        parameters.forEach { p ->
            if (p != firstPrm) { //最初の要素以外に対して
                val currentBvs = boundaryValueList[p]

                //inputDataListの第一引数のみを登録した状態
                val inputDataListInitialState = ArrayList(inputDataList)

                for (i in 0 until currentBvs!!.size - 1) {
                    val inputDataListTmp = InputDataList()
                    inputDataListInitialState.forEach { inputDataOriginal ->
                        //inputDataを複製
                        val inputData = HashMap<String, Long>()
                        inputDataOriginal.forEach({ key, value -> inputData[key] = value })
                        inputDataListTmp.add(inputData)
                    }
                    inputDataList.addAll(inputDataListTmp)
                }
                for ((repeatTimesOfInsert, current_bv) in currentBvs.withIndex()) {
                    val offset = repeatTimesOfInsert * inputDataListInitialState.size
                    for (k in inputDataListInitialState.indices) {
                        val inputData = inputDataList[k + offset]
                        inputData[p] = current_bv
                    }
                }

            }
        }
    } //end makeInputDatList

    companion object {
        internal const val intMax = Integer.MAX_VALUE.toLong()
        internal const val intMin = Integer.MIN_VALUE.toLong()
        internal const val natMax = intMax * 2
        internal const val natMin: Long = 0
        internal const val nat1Max = natMax + 1
        internal const val nat1Min: Long = 1
    }

}
